File: DynamicPrograming/csharp/ValidCompound.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;

namespace DynamicProgrammingSolutions;

public static class ValidCompound
{
    public static bool IsValid(string compound, IEnumerable<string> elements)
    {
        var lowerCompound = compound.ToLower(CultureInfo.InvariantCulture);
        var lowerElements = elements.Select(e => e.ToLower(CultureInfo.InvariantCulture)).ToArray();
        var memo = new Dictionary<int, bool>();
        return Dfs(0, lowerCompound, lowerElements, memo);
    }

    private static bool Dfs(int index, string compound, IReadOnlyList<string> elements, IDictionary<int, bool> memo)
    {
        if (index == compound.Length)
        {
            return true;
        }

        if (memo.TryGetValue(index, out var cached))
        {
            return cached;
        }

        foreach (var element in elements)
        {
            if (index + element.Length <= compound.Length &&
                compound.AsSpan(index, element.Length).SequenceEqual(element.AsSpan()) &&
                Dfs(index + element.Length, compound, elements, memo))
            {
                memo[index] = true;
                return true;
            }
        }

        memo[index] = false;
        return false;
    }
}

File: DynamicPrograming/go/valid_compound.go
package dynamicprogramming

import "strings"

// ValidCompound returns true if compound can be spelled using element symbols.
func ValidCompound(compound string, elements []string) bool {
	lowerCompound := strings.ToLower(compound)
	lowerElements := make([]string, len(elements))
	for i, el := range elements {
		lowerElements[i] = strings.ToLower(el)
	}

	memo := make(map[int]bool)

	var dfs func(int) bool
	dfs = func(i int) bool {
		if i == len(lowerCompound) {
			return true
		}
		if val, ok := memo[i]; ok {
			return val
		}
		for _, element := range lowerElements {
			if i+len(element) <= len(lowerCompound) && lowerCompound[i:i+len(element)] == element {
				if dfs(i + len(element)) {
					memo[i] = true
					return true
				}
			}
		}
		memo[i] = false
		return false
	}

	return dfs(0)
}
